File: agent/src/runtime/events.ts
/** JSON 能安全传输的数据类型，与前端 IM 的 JsonValue 含义一致。 */
export type JsonValue =
  | string
  | number
  | boolean
  | null
  | JsonValue[]
  | { [key: string]: JsonValue };

/**
 * Node、LangChain/Deep Agents 都不能改变这组领域事件。
 * Go 后续只需要把它们映射为前端 IM envelope。
 */
export type AgentEventType =
  | 'run.started'
  | 'run.completed'
  | 'run.failed'
  | 'thinking.delta'
  | 'thinking.completed'
  | 'content.started'
  | 'content.delta'
  | 'content.completed'
  | 'tool.started'
  | 'tool.completed'
  | 'tool.failed';

export interface AgentEvent {
  type: AgentEventType;
  runId: string;
  payload: Record<string, JsonValue>;
}

export interface AgentError {
  code: string;
  message: string;
  retryable: boolean;
}

export function runFailed(
  runId: string,
  code: string,
  message: string,
  retryable: boolean,
): AgentEvent {
  return {
    type: 'run.failed',
    runId,
    payload: {
      error: { code, message, retryable },
    },
  };
}

File: agent/eterion_agent/runtime/events.py
"""Normalized events emitted by every Agent runtime.

These events describe Agent execution rather than a transport envelope. The Go
adapter can add thread sequence numbers, timestamps, and message identifiers
without knowing anything about LangChain, DeepAgents, or provider chunks.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Literal, TypeAlias


JsonValue: TypeAlias = (
    str | int | float | bool | None | list["JsonValue"] | dict[str, "JsonValue"]
)

AgentEventType: TypeAlias = Literal[
    "run.started",
    "run.completed",
    "run.failed",
    "thinking.delta",
    "thinking.completed",
    "content.started",
    "content.delta",
    "content.completed",
    "tool.started",
    "tool.completed",
    "tool.failed",
]


@dataclass(frozen=True, slots=True)
class AgentEvent:
    type: AgentEventType
    run_id: str
    payload: dict[str, JsonValue]


def failure(run_id: str, code: str, message: str, retryable: bool) -> AgentEvent:
    return AgentEvent(
        "run.failed",
        run_id,
        {
            "error": {
                "code": code,
                "message": message,
                "retryable": retryable,
            }
        },
    )
